pontoon/base/models.py

```diff
@@ -1675,6 +1687,87 @@ class ChangedEntityLocale(models.Model):
 
     class Meta:
         unique_together = ('entity', 'locale')
+        
+
+class EntityFiltersManager(models.Manager):
+    def update_filters_state(self, entity, locale):
+        """
+        Update state of filters of entity for a given locale.
+        """
+        filters_instance = self.get_queryset().get(entity=entity)
+
+        forms_count = locale.nplurals if entity.string_plural else 1
+        translations = entity.translation_set.filter(locale=locale).values_list('string', 'approved', 'fuzzy')
+
+        def set_field(field_name, value):
+            setattr(filters_instance, '{}_{}'.format(locale.code, field_name), value)
+
+        set_field('status', self.get_entity_status(translations, locale))
+        set_field('unchanged', self.is_unchanged(entity, translations, forms_count))
+        set_field('has_suggestions', self.has_suggestions(translations))
+        filters_instance.save()
+
+    def get_entity_status(self, translations, forms_count):
+        """
+        Calculate current status of an entity.
+        """
+
+        if forms_count == len([1 for _, approved, _ in translations if approved]):
+            return 'translated'
+
+        elif forms_count == len([1 for _, _, fuzzy in translations if fuzzy]):
+            return 'fuzzy'
+
+        elif len([1 for _, approved, fuzzy in translations if not approved and not fuzzy]):
+            return 'suggested'
+
+        return 'missing'
+
+    def is_unchanged(self, entity, translations, forms_count):
+        """
+        Checks if entity has been changed.
+        """
+        strings = [s for s, _, _ in translations]
+        same_string = entity.string in strings
+        if same_string and forms_count > 1:
+            return entity.string_plural in strings
+        return same_string
+
+    def has_suggestions(self, translations):
+        return bool(len([1 for _, approved, fuzzy in translations if not approved and not fuzzy]))
+
+
+class EntityFilters(models.Model):
```

Comment:
I'm not sure if it would be significantly slower but I think that design would require more DB operations after every change in locale/entity (e.g. When you add/remove locale from the project, that would cost a lot of deletes/inserts for every locale). Probably, that could make initial migration even slower.